### PR TITLE
[BEAM-570] Performance optimization in fileio for compressed sources.

### PR DESCRIPTION
**Pull request by katsiapis**
- Moving from a string-based buffer to a cStringIO based on in order to help with performance of compressed sources.
- Dropping some TODOs for doing similar performance improvements in GcsIO.
- Making _CompressedFile conform to standard Python IO function names.

### CODE REVIEW

**Comment by katsiapis** [comment]
Hi @chamikaramj, can you please take a look?
R: @chamikaramj

**Comment by chamikaramj** [comment]
LGTM

**Comment by katsiapis** [comment]
Hi @robertwb, could you merge this in?
R: @robertwb

File: sdks/python/apache_beam/io/fileio.py [hunk 1]
```diff
@@ -708,10 +708,26 @@ def write(self, data):
...
-    while not self._read_eof and len(self._data) < num_bytes:
+    if (not self._read_eof and self._read_position > 0 and
+        (self._read_buffer.tell() - self._read_position) < num_bytes):
+      # There aren't enough number of bytes to accommodate a read, so we
+      # prepare for a possibly large read by clearing up all internal buffers
+      # but without dropping any previous held data.
+      self._read_buffer.seek(self._read_position)
+      data = self._read_buffer.read()
+      self._read_position = 0
+      self._read_buffer.truncate(0)
```
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: @robertwb , should we also need self._read_buffer.seek(0) here for "completeness"?

Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: I'm not sure if truncate does this (it doesn't for ordinary files), so probably we should.

Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Done.

**Comment by katsiapis** [comment]
Thanks for the merge @robertwb! Closing...

**Pull request closed without merging**